### PR TITLE
#59 used parameters

### PR DESCRIPTION
**Pull request by Yuri05**

### CODE REVIEW

File: src/OSPSuite.SimModel/include/SimModel/SimulationTask.h [hunk 1]
```diff
@@ -29,6 +29,9 @@ class SimulationTask
 	static void SetValuesBelowAbsTolLevelToZero(double * values, int valuesSize, double absTol);
 
 	static void CheckForNegativeValues(Species ** odeVariables, int numberOfVariables, double absTol);
+
+	//mark all parameters which are used in any ODE varaible or observer as unused
+	static void MarkUsedParameters(Simulation * sim);
```
Comment (Yuri05) [inline src/OSPSuite.SimModel/include/SimModel/SimulationTask.h]: That's the main routine, which is called between Simulation.Load and Simulation.Finalize.
It sets `UsedInSimulation ` flag for every simulation parameter

File: src/OSPSuite.SimModel/src/SimulationOptions.cpp [hunk 2]
```diff
@@ -15,13 +15,14 @@ using namespace std;
 
 SimulationOptions::SimulationOptions(void)
 {
-	_validateXMLWithSchema = true;
+	_validateXMLWithSchema = false; //perform schema validation only if explicitly set by user
```
Comment (Yuri05) [inline src/OSPSuite.SimModel/src/SimulationOptions.cpp]: this has nothing to do with used parameters.
Just changed the default from true (="validate XML with SimModel-Schema") to false. Just to increase the performance.
PK-Sim/MoBi always set it to false (assuming models generated there are always valid :)
Matlab/R-Toolbox still can set it to true if required

File: src/OSPSuite.SimModel/src/SimulationOptions.cpp [hunk 3]
```diff
@@ -15,13 +15,14 @@ using namespace std;
...
 	_showProgress = false;
 	_executionTimeLimit = 0.0;    //no limit of simulation execution time
 	_stopOnWarnings = true;       //treat warnings of ODE solver as errors
 	_autoReduceTolerances = true; //automatically reduce solver tolerances
 	_writeLogFile = true;
 	_checkForNegativeValues = true;
+	_identifyUsedParameters = false;
```
Comment (Yuri05) [inline src/OSPSuite.SimModel/src/SimulationOptions.cpp]: Not sure if we need the flag or if used parameters should be ALWAYS estimated.
Currently there is a posiibility to skip this step (in this case, ALL parameters will be set as used)

File: src/OSPSuite.SimModel/src/SimulationTask.cpp [hunk 4]
```diff
@@ -222,6 +222,54 @@ void SimulationTask::CheckForNegativeValues(Species ** odeVariables, int numberO
 
 }
 
+void SimulationTask::MarkUsedParameters(Simulation * sim)
+{
+	TObjectList<Parameter> & parameters = sim->Parameters();
+	int idx;
+
+	if (!sim->Options().IdentifyUsedParameters())
```
Comment (Yuri05) [inline src/OSPSuite.SimModel/src/SimulationTask.cpp]: s. comment regarding this option above.

File: src/OSPSuite.SimModel/src/SimulationTask.cpp [hunk 5]
```diff
@@ -222,6 +222,54 @@ void SimulationTask::CheckForNegativeValues(Species ** odeVariables, int numberO
 
 }
 
+void SimulationTask::MarkUsedParameters(Simulation * sim)
```
Comment (Yuri05) [inline src/OSPSuite.SimModel/src/SimulationTask.cpp]: Main routine. MArks all parameters as used, which are used in:
* ODE Variables (RHS or initial formula)
* Observers
* Event conditions
* Event assignment formulas

File: src/OSPSuite.SimModel/src/Species.cpp [hunk 6]
```diff
@@ -448,9 +448,14 @@ void Species::SetODEIndex(int newIndex)
 	m_ODEIndex = newIndex;
 }
 
-void Species::AppendUsedParameters(std::set<int> & usedParameterIDs) {
+void Species::AppendUsedParameters(std::set<int> & usedParameterIDs, bool includeInitialFormula)
```
Comment (Yuri05) [inline src/OSPSuite.SimModel/src/Species.cpp]: @hdiedam Not sure why initial formula is skipped when getting used parameters for CPP-Export.
Thus I added optional parameter includeInitialFormula (false per default, so CPP-Export is not affected)

File: src/OSPSuite.SimModel/src/Switch.cpp [hunk 7]
```diff
@@ -178,14 +178,14 @@ void Switch::AppendUsedVariables(set<int> & usedVariblesIndices)
 		_formulaChangeVector[i]->AppendVariablesUsedInNewFormula(usedVariblesIndices, emptySet);
 }
 
-void Switch::AppendUsedParameters(std::set<int> & usedParameterIDs)
+void Switch::AppendUsedParameters(std::set<int> & usedParameterIDs, bool alwaysAppendInFormulaChange)
```
Comment (Yuri05) [inline src/OSPSuite.SimModel/src/Switch.cpp]: s. comment in FormulaChange above

File: src/OSPSuite.SimModelComp/OSPSuite_SimModelComp.xml [hunk 8]
```diff
@@ -21,6 +21,14 @@
 						<FieldDef Name="ExecutionTimeLimit" DataType="Double" Description="Execution time limit for single simulation run (0= no limit)" DefaultValue="0"/>
 						<Values>0</Values>
 					</Variable>
+					<Variable Key="ValidateWithXMLSchema">
```
Comment (Yuri05) [inline src/OSPSuite.SimModelComp/OSPSuite_SimModelComp.xml]: Configuration file for Matlab/R-Toolbox extend with possibilities to
* enable/disable validation with SimModel-Schema
* enable/disable identifying used parameters (if false: ALL parameters will be marked as used)

File: src/OSPSuite.SimModelComp/include/SimModelComp/SimModelComp.h [hunk 9]
```diff
@@ -61,6 +61,9 @@ class SimModelComp : public DCI::Component
 	bool m_ProcessDataFirstRun;
 
 	void LoadConfigurationFromParameterTable();
+	bool GetBooleanValueFrom(DCI::ITableHandle & hTab, const std::string & columnName, bool defaultValue);
+	double GetDoubleValueFrom(DCI::ITableHandle & hTab, const std::string & columnName, double defaultValue);
+	std::string GetStringValueFrom(DCI::ITableHandle & hTab, const std::string & columnName, const std::string & defaultValue="");
```
Comment (Yuri05) [inline src/OSPSuite.SimModelComp/include/SimModelComp/SimModelComp.h]: just some refactoring

File: tests/TestData/TestExportUsedParameters02.xml [hunk 10]
```diff
@@ -0,0 +1,225 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (Yuri05) [inline tests/TestData/TestExportUsedParameters02.xml]: Simulation used in tests, exported from TestExportUsedParameters01.mbp3

File: src/OSPSuite.SimModel/include/SimModel/SimulationOptions.h [hunk 11]
```diff
@@ -22,6 +22,8 @@ namespace SimModelNative
 		bool _keepXMLNodeAsString; //original xml is required only for saving the simulation to XML
 		bool _useFloatComparisonInUserOutputTimePoints; //if set to true, float comparison will be used
 		                                                //for user output time points.Otherwise: double
+		bool _identifyUsedParameters; //if set to false: ALL parameters will be marked as used in ODE variables/observes
```
Comment (msevestre) [inline src/OSPSuite.SimModel/include/SimModel/SimulationOptions.h]: is there a reason why comments are not above the variable? Kind of unusual

File: src/OSPSuite.SimModel/include/SimModel/FormulaChange.h [hunk 12]
```diff
@@ -55,7 +55,7 @@ class FormulaChange :
 
 	//used variables are only added into the list if UseAsValue=false
 	void AppendVariablesUsedInNewFormula(std::set<int> & usedVariblesIndices, const std::set<int> & variblesIndicesUsedInSwitchAssignments);
-	void AppendUsedParameters(std::set<int> & usedParameterIDs);
+	void AppendUsedParameters(std::set<int> & usedParameterIDs, bool alwaysAppend);
```
Comment (msevestre) [inline src/OSPSuite.SimModel/include/SimModel/FormulaChange.h]: don't understand what this parameter means `alwaysAppend`

File: src/OSPSuite.SimModel/include/SimModel/Species.h [hunk 13]
```diff
@@ -100,7 +100,7 @@ class Species :
 
 	void SetODEIndex(int newIndex);
 
-	virtual void AppendUsedParameters(std::set<int> & usedParameterIDs);
+	virtual void AppendUsedParameters(std::set<int> & usedParameterIDs, bool includeInitialFormula=false);
```
Comment (msevestre) [inline src/OSPSuite.SimModel/include/SimModel/Species.h]: What does this parameter mean `includeInitialFormula`
Sometime this extra parameter has a default value (here false) sometimes not (see FormulaChange above). Why?

File: src/OSPSuite.SimModel/include/SimModel/SimulationOptions.h (hunk 11 above)
Comment (Yuri05) [reply src/OSPSuite.SimModel/include/SimModel/SimulationOptions.h]: No arguable reason :)

File: src/OSPSuite.SimModel/include/SimModel/FormulaChange.h (hunk 12 above)
Comment (Yuri05) [reply src/OSPSuite.SimModel/include/SimModel/FormulaChange.h]: Is explained here: https://github.com/Open-Systems-Pharmacology/OSPSuite.SimModel/pull/60#pullrequestreview-200233305

File: src/OSPSuite.SimModel/include/SimModel/Species.h (hunk 13 above)
Comment (Yuri05) [reply src/OSPSuite.SimModel/include/SimModel/Species.h]: Is explained here: https://github.com/Open-Systems-Pharmacology/OSPSuite.SimModel/pull/60#pullrequestreview-200236987
The reason is: `Species.AppendUsedParameters` is called "on the top level" 
a) from the current use case and 
b) from the Model-CPP-Export). 
As I didn't want to change b), I defined this parameter as optional.

`FormulaChange.AppendUsedParameters` is called only by it's "parent" (`Switch.AppendUsedParameters`), thus here there is no need for optional paramter

File: src/OSPSuite.SimModel/src/Species.cpp (hunk 6 above)
Comment (hdiedam) [reply src/OSPSuite.SimModel/src/Species.cpp]: The cppExport uses that function primarily to distinguish between "local" parameters (switched or somehow depending on y and t) and "global" parameters (depending only on free parameters) that can be precomputed. If they are depending only on some initial value y(t_0) and not on the state trajectory y(t) they still can be considered global.

**Comment by Yuri05** [comment]
Merging to continue.
@hdiedam you can still review

**Pull request merged**